tt_metal/impl/dispatch/kernels/cq_dispatch.cpp

```diff
@@ -125,6 +141,123 @@ void get_dispatch_cb_page() {
+    uint32_t completion_wr_ptr_and_toggle = cq_write_interface.completion_fifo_wr_ptr | (cq_write_interface.completion_fifo_wr_toggle << 31);
+    volatile tt_l1_ptr uint32_t* completion_wr_ptr_addr = get_cq_completion_write_ptr();
+    completion_wr_ptr_addr[0] = completion_wr_ptr_and_toggle;
+    noc_async_write(CQ_COMPLETION_WRITE_PTR, pcie_address, 4);
+    block_noc_writes_to_clear[rd_block_idx]++;
+}
+
+FORCE_INLINE
+void completion_queue_push_back(uint32_t num_pages) {
+    // Transfer pages are aligned
+    uint32_t push_size_16B = num_pages * completion_queue_page_size_16B;
+    cq_write_interface.completion_fifo_wr_ptr += push_size_16B;
+
+    if (cq_write_interface.completion_fifo_wr_ptr >= completion_queue_end_addr_16B) {
+        cq_write_interface.completion_fifo_wr_ptr = cq_write_interface.completion_fifo_wr_ptr - completion_queue_end_addr_16B + completion_queue_base_addr_16B;
+        // Flip the toggle
+        cq_write_interface.completion_fifo_wr_toggle = not cq_write_interface.completion_fifo_wr_toggle;
+    }
+
+    // Notify host of updated completion wr ptr
+    notify_host_of_completion_queue_write_pointer();
+}
+
+
+FORCE_INLINE
+void process_write_host() {
+    volatile tt_l1_ptr CQDispatchCmd *cmd = (volatile tt_l1_ptr CQDispatchCmd *)cmd_ptr;
+
+    uint32_t completion_write_ptr;
+    // We will send the cmd back in the first X bytes, this makes the logic of reserving/pushing completion queue
+    // pages much simpler since we are always sending writing full pages (except for last page)
+    uint32_t length = cmd->write.length;
+    uint32_t data_ptr = cmd_ptr;
+    while (length != 0) {
+        // Get a page if needed
+        if (cb_fence == data_ptr) {
+            // Check for block completion
+            if (cb_fence == block_next_start_addr[rd_block_idx]) {
+                // Check for dispatch_cb wrap
+                if (rd_block_idx == dispatch_cb_blocks - 1) {
+                    cb_fence = dispatch_cb_base;
+                    data_ptr = dispatch_cb_base;
+                }
+                move_rd_to_next_block();
+            }
+            // Wait for dispatcher to supply a page (this won't go beyond the buffer end)
+            uint32_t n_pages = dispatch_cb_acquire_pages();
+            cb_fence += n_pages * dispatch_cb_page_size;
+
+            // Release pages for prefetcher
+            // Since we gate how much we acquire to < 1/4 the buffer, this should be called enough
+            dispatch_cb_block_release_pages();
+        }
+        uint32_t available_data = cb_fence - data_ptr;
+        uint32_t xfer_size = (length > available_data) ? available_data : length;
+        uint32_t npages = (xfer_size + completion_queue_page_size - 1) / completion_queue_page_size;
+        completion_queue_reserve_back(npages);
+        uint32_t completion_queue_write_addr = cq_write_interface.completion_fifo_wr_ptr << 4;
```

Comment:
<< 4 ?  is one of these 16B?

Reply:
Yes, cq_write_interface.completion_fifo_wr_ptr is stored as address >> 4 like FD 1